app/Observers/RemindablePost.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+namespace App\Observers;
+
+use App\Models\Calendar;
+use App\Models\Entity;
+use App\Models\EntityEventType;
+use App\Models\Post;
+use App\Models\Quest;
+use App\Models\Reminder;
+use Exception;
+use Illuminate\Database\Eloquent\Model;
+
+class RemindablePost
+{
+    public function saved(Post $post)
+    {
+        // If we don't have an entity, not exactly sure what's going on. Skip the entity event
+        // observer and let the user report it instead of throwing an ugly error at them.
+        if (empty($post->entity)) {
+            return;
+        }
+
+        // The user is editing an entity with a calendar, but doesn't have the permission to see
+        // the calendar? We skip any work.
+        if (request()->has('calendar_skip')) {
+            return;
+        }
+
+        // Previously, this lookup was only triggered when the calendar_id or date was dirty. However, this excludes just
+        // changing the colour or periodicity. To support the API not overriding the values, we still check to make
+        // sure that the calendar_id property is set.
+        if (! request()->has('calendar_id')) {
+            return;
+        }
+        
+        $calendarID = request()->post('calendar_id');
+
+        // We already had this event linked
+        /** @var Quest $post */
+        $reminder = $post->calendarReminder();
+        if ($reminder !== null) {
+            // We no longer have a calendar attached to this model
+            if ($calendarID === null) {
+                $reminder->delete();
+
+                return;
+            }
+        } else {
+            $reminder = new Reminder;
+            $reminder->remindable_type = Post::class;
+            $reminder->remindable_id = $post->id;
+        }
+
+        // Validate the calendar
+        /** @var ?Calendar $calendar */
+        $calendar = Calendar::find($calendarID);
+        if ($calendar === null || $calendar->missingDetails()) {
+            return;
+        }
+
+        $length = request()->post('calendar_length', '1');
```

Comment:
So much copy-pasted code. Move this all to a service and have all observers call the service instead